Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm

```diff
@@ -0,0 +1,301 @@
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "LibraryPathDiagnostics.h"
+
+#include <dlfcn.h>
+#include <notify.h>
+#include <os/log.h>
+#include <wtf/FileSystem.h>
+#include <wtf/JSONValues.h>
+#include <wtf/Span.h>
+#include <wtf/StringPrintStream.h>
+#include <wtf/UUID.h>
+#include <wtf/cf/TypeCastsCF.h>
+#include <wtf/spi/cf/CFPrivSPI.h>
+#include <wtf/spi/darwin/dyldSPI.h>
+
+#if HAVE(CHIRP_SPI)
+// Workaround for rdar://97945223
+#define dlsym (const canary_meta_t *)dlsym
+#include <chirp/chirp.h>
+#undef dlsym
+#endif
+
+#if HAVE(SHARED_REGION_SPI)
+#include <mach/shared_region.h>
+#endif
+
+namespace WTF {
+
+static char const * const notificationName = "com.apple.WebKit.LibraryPathDiagnostics";
+static char const * const loggingSubsystem = "com.apple.WebKit.LibraryPathDiagnostics";
+static char const * const loggingCategory = "LibraryPathDiagnostics";
+#if HAVE(DYLD_DLOPEN_IMAGE_HEADER_SPI)
+static char const * const libraryListEnvironmentVariableName = "LIBRARY_PATH_DIAGNOSTICS_LIBRARIES";
+#endif
+static char const * const bundleListEnvironmentVariableName = "LIBRARY_PATH_DIAGNOSTICS_BUNDLES";
+
+class LibraryPathDiagnosticsLogger final {
+public:
+    LibraryPathDiagnosticsLogger();
+    void log(void);
+private:
+    void logJSONPayload(const JSON::Object&);
+    void logString(const Vector<String>& path, const String&);
+    void logObject(const Vector<String>& path, Ref<JSON::Object>&&);
+    void logError(const char*, ...);
+
+    void logExecutablePath(void);
+    void logDYLDSharedCacheInfo(void);
+#if HAVE(DYLD_DLOPEN_IMAGE_HEADER_SPI)
+    void logDynamicLibraryInfo(const String&);
+    void logDynamicLibraryInfo(void);
+#endif
+    void logBundleInfo(const String&);
+    void logBundleInfo(void);
+#if HAVE(CHIRP_SPI)
+    void logCryptexCanaryInfo(canary_cryptex_t, const String&);
+    void logCryptexCanaryInfo(void);
+#endif
+
+    os_log_t m_osLog;
+};
+
+LibraryPathDiagnosticsLogger::LibraryPathDiagnosticsLogger()
+    : m_osLog(os_log_create(loggingSubsystem, loggingCategory))
+{
+}
+
+void LibraryPathDiagnosticsLogger::logJSONPayload(const JSON::Object &object)
+{
+    auto textRepresentation = object.toJSONString();
+    os_log(m_osLog, "%{public}s", textRepresentation.utf8().data());
+}
+
+void LibraryPathDiagnosticsLogger::logString(const Vector<String>& path, const String& string)
+{
+    auto payload = JSON::Object::create();
+    auto iter = path.rbegin();
+    payload->setString(*iter, string);
+    for (iter++; iter != path.rend(); iter++) {
```

Comment:
Can we use modern `for (:)` iteration here?

Reply:
Oh I see, you're using the iterator once before the loop. This is a nit, but I probably would've used an indexed loop instead of this.